### ISSUE TITLE
@CartesianProductTest - ParameterResolutionException with test class constructor with single parameter for TestInfo

### ISSUE THREAD

**Issue by richardwalsh** [junit-pioneer/junit-pioneer#401]
I am updating a bunch tests to use @CartesianProductTest and ran into parameter resolution error for the test class constructor:

`
org.junit.jupiter.api.extension.ParameterResolutionException: Discovered multiple competing ParameterResolvers for parameter [org.junit.jupiter.api.TestInfo arg0] in constructor [public com.blah.SillyTest(org.junit.jupiter.api.TestInfo)]: org.junit.jupiter.engine.extension.TestInfoParameterResolver@8646db9, org.junitpioneer.jupiter.CartesianProductResolver@37374a5e
`

```java
import org.junit.jupiter.api.TestInfo;
import org.junitpioneer.jupiter.CartesianProductTest;
import org.junitpioneer.jupiter.CartesianValueSource;

public class SillyTest {

    private TestInfo testInfo;

    public SillyTest(TestInfo testInfo) {
        this.testInfo = testInfo;
    }

    @CartesianProductTest
    @CartesianValueSource(ints = {1, 2})
    @CartesianValueSource(ints = {3, 4})
    void myCartesianTestMethod(int x, int y) {
        System.out.println("x: " + x + ", y: " + y);
    }
}
```

The suggestion on the #junit-pioneer discord (thanks!) was to remove TestInfo from the constructor and add it as the last parameter and it worked:

```java
import org.junit.jupiter.api.TestInfo;
import org.junitpioneer.jupiter.CartesianProductTest;
import org.junitpioneer.jupiter.CartesianValueSource;

public class SillyTest2 {

    @CartesianProductTest
    @CartesianValueSource(ints = {1, 2})
    @CartesianValueSource(ints = {3, 4})
    void myCartesianTestMethod(int x, int y, TestInfo testInfo) {
        System.out.println("x: " + x + ", y: " + y);
    }
}
```

Is the constructor approach possible with pioneer?

**Comment by Michael1993** [comment]
This might take a while to properly investigate/fix.

There are some major changes coming to how CartesianProductTest works (internally). [See the related PR](https://github.com/junit-pioneer/junit-pioneer/pull/362).

**Comment by Michael1993** [comment]
I looked through the issue and identified the cause. We can fix this by making the extension more 'diligent' about what it should and should not resolve.

In more detail: `CartesianProductResolver#supportsParameter` method must be updated with more logic.
Currently it's just an index check, should check type as well. This could mean that we need to pass more data to `CartesianProductResolver`, to know what we are supposed to resolve - or we just exclude `TestInfo`, `TestReporter`, etc. The latter is a **lot** simpler but it's also fragile and does not work with other, "non-Cartesian" parameters.

**Comment by Michael1993** [comment]
So, after being anxious about this issue being *very hard* at 2 AM, it turns out it's a fairly straightforward fix (unless I'm missing something).

**Comment by richardwalsh** [comment]
Very cool, thanks for fixing this!

**Comment by richardwalsh** [comment]
The @BeforeEach method has a similar issue with respect to "Discovered multiple competing ParameterResolvers for parameter [org.junit.jupiter.api.TestInfo arg0]". Will this fix address that too?

**Comment by Michael1993** [comment]
Yes, that is correct. After the fix, it will only support test methods.

**Comment by nipafx** [comment]
@richardwalsh We just released 1.2.0 that contains this fix. Let us know if any problems remain. 😃

**Comment by richardwalsh** [comment]
> @richardwalsh We just released 1.2.0 that contains this fix. Let us know if any problems remain. 😃

Worked perfectly, thanks!
